examples/addr2line.rs

```diff
@@ -111,29 +69,87 @@ fn symbolicate<Endian>(file: &object::File, matches: &getopts::Matches)
         .expect("Can't addr2line without .debug_line");
     let debug_line = gimli::DebugLine::<Endian>::new(&debug_line);
 
-    for (info_offset, addr) in offsets.iter().zip(addrs.iter()) {
-        match *info_offset {
-            None => println!("Found nothing"),
-            Some(d) => {
-                match debug_info.header_from_offset(d) {
-                    Err(_) => println!("Couldn't get DIE header"),
-                    Ok(h) => {
-                        let line_offset = line_offset_for_entry(&debug_abbrev, &h)
-                            .expect("No offset into .debug_lines!?");
-                        let header = gimli::LineNumberProgramHeader::new(debug_line,
-                                                                         line_offset,
-                                                                         h.address_size());
-                        if let Ok(header) = header {
-                            let mut state_machine = gimli::StateMachine::new(header);
-                            match state_machine.run_to_address(addr) {
-                                Err(_) => println!("Failed to run line number program!"),
-                                Ok(None) => println!("Failed to find matching line for {}", *addr),
-                                Ok(Some(row)) => display_file(row),
-                            };
-                        }
-                    }
-                }
+    let mut units = Vec::new();
+    let mut headers = debug_info.units();
+    while let Some(header) = headers.next().expect("Couldn't get DIE header") {
+        if let Some(unit) = Unit::parse(&debug_abbrev, &header) {
+            units.push(unit);
+        }
+    }
+
+    for addr in addrs {
+        find_address(debug_line, &units, addr);
+    }
+}
+
+fn find_address<Endian>(debug_line: gimli::DebugLine<Endian>, units: &[Unit], addr: u64)
+    where Endian: gimli::Endianity
+{
+    for unit in units {
+        if unit.contains_address(addr) {
```

Comment:
Possible follow up: investigate sorting units by their addresses and doing a binary search here.